---
title: "Alzheimer_Brain_GSE134379"
author: "Aitana Egea"
date: "2024-04-18"
output: html_document
---

```{r setup, include=FALSE}
# Load necessary libraries
library(knitr)
library(tidyverse)
library(limma)
library(minfi)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
library(IlluminaHumanMethylation450kmanifest)
library(RColorBrewer)
library(missMethyl)
library(minfiData)
library(Gviz)
library(DMRcate)
library(stringr)
library(chrXRa)
library(GEOquery)
library(data.table)

```

```{r}

gsm.expr <- getGEO("GSE134379", destdir = ".", AnnotGPL = TRUE)[[1]]

```

```{r}
pheno <- pData(gsm.expr)
pheno <- pheno %>% filter(`gender:ch1` == "F")

```


# PIPELINE FUNCTIONS. 
```{r}

# Función para leer los nombres de los archivos .idat de una carpeta
listIdatFiles <- function(folderPath) {
  # Obtener todos los nombres de archivos .idat y .idat.gz en la carpeta
  files <- list.files(folderPath, pattern = "\\.idat(\\.gz)?$", full.names = FALSE)
  return(files)
}


addBasenameColumn <- function(targets, files, basePath) {
  # Modificar la columna Basename para que contenga la parte base del nombre de archivo
  targets$Basename <- sapply(targets$geo_accession, function(accession) {
    # Encontrar archivos que coincidan con la accession en su nombre
    matching_files <- grep(accession, files, value = TRUE)
    
    # Asumir que los nombres de los archivos contienen la accession seguida de _Grn o _Red
    # Extraer la parte base común del nombre de archivo, excluyendo el sufijo y la extensión
    # Esta operación asume que todos los archivos para una muestra tienen el mismo prefijo
    if(length(matching_files) > 0) {
      basename_common <- sub("(_Grn|_Red)\\.idat(\\.gz)?$", "", matching_files[1])
      return(paste0(basePath, "/", basename_common))
    }
    
    return(NA)
  })
  
  return(targets)
}


preprocess_rgset <- function(rgSet, targets) {
  
  detP <- detectionP(rgSet)
  # remove poor quality samples
  keep <- colMeans(detP) < 0.05
  
  # remove poor quality samples
  rgSet <- rgSet[,keep]
  
  # remove poor quality samples from targets data
  targets <- targets[keep,]
  
  # remove poor quality samples from detection p-value table
  detP <- detP[,keep]
  
  # normalize the data; this results in a GenomicRatioSet object
  mSetSq <- preprocessQuantile(rgSet) 
  
  #Filtering: 
  detP <- detP[match(featureNames(mSetSq),rownames(detP)),] 

  # remove any probes that have failed in one or more samples
  keep <- rowSums(detP < 0.01) == ncol(mSetSq)
  mSetSqFlt <- mSetSq[keep,]
  mSetSqFlt <- dropLociWithSnps(mSetSqFlt)
 
  
 # C:\Users\aegea\Downloads\48639-non-specific-probes-Illumina450k.csv
xReactiveProbes <- read.csv("C:/Users/aegea/Downloads/48639-non-specific-probes-Illumina450k.csv")
  keep <- !(featureNames(mSetSqFlt) %in% xReactiveProbes$TargetID)
  mSetSqFlt <- mSetSqFlt[keep,] 
  mVals <- getM(mSetSqFlt)
  bVals <- getBeta(mSetSqFlt)
  return(list(mVals = mVals, bVals = bVals))
}
```



```{r}
# Specify the path to the tar file and the target directory for extraction
tarFilePath <- "C:/Users/aegea/Downloads/GSE134379_RAW.tar"
targetDirectory <- "C:/Users/aegea/Documents/TFM/GSE134379/idat"

# Create the target directory if it doesn't exist
if (!dir.exists(targetDirectory)) {
  dir.create(targetDirectory, recursive = TRUE)
}

# Extract the tar file
untar(tarFilePath, exdir = targetDirectory)
```


```{r}
files <- listIdatFiles("C:/Users/aegea/Documents/TFM/GSE134379/idat")
targets <-addBasenameColumn(pheno, files, "C:/Users/aegea/Documents/TFM/GSE134379/idat")
rgset<- read.metharray.exp(base="C:/Users/aegea/Documents/TFM/GSE134379/idat", targets=targets)
bvals<- preprocess_rgset(rgset, targets)$bVals

```


3. Load the annotation data from cpgid.RData file and calculate X-Ra biomarker

```{r}
load("C:/Users/aegea/Documents/TFM/ALZHEIMER/Repo_Alzheimer_Master_Thesis/Supplementary_Files/cpgid.RData")

# Define the function for X-Ra calculation
calculate_xra <- function(df, met, title, output_prefix){
  
  gsms <- df %>% pull(geo_accession)
  cpgx <- rownames(cpgid)[which(cpgid$Chromosome=="X")]

  met_subset <- met[, gsms]
  
  common_CpGs <- intersect(rownames(met_subset), cpgx)
  bvals_chrX <- met_subset[common_CpGs, ]
  metfemale <- t(bvals_chrX)
  df$XRa <- XRa(metfemale, colnames(metfemale))
 
  return(df)
}

```



5. Fit and store Generalize Linear Models for X-Ra's effect 
on disease status for each brain region

```{r}


colnames(bvals)<- gsub("_(.*)", "", colnames(bvals)) 
pheno_xra <- calculate_xra(pheno, bvals) %>%
  mutate(
    brain_region = `brain region:ch1`, 
    age = as.numeric(`age:ch1`),
    disease_status = ifelse(`diagnosis:ch1` == "AD", 1,0),
    xra_100 = XRa * 100
 
  )

# Function to create and save GLM models for different tissue types
create_glm_models <- function(pheno, dir_path, accession_name) {
  # Ensure the directory exists or create it
  if (!dir.exists(dir_path)) {
    dir.create(dir_path, recursive = TRUE)
    cat("Directory created at:", dir_path, "\n")
  } else {
    cat("Directory already exists at:", dir_path, "\n")
  }

  # Loop through each unique tissue type
  unique(pheno$brain_region) %>%
    set_names() %>%
    map(~{
      # Filter the data for the specific tissue
      data_subset <- filter(pheno, brain_region == .x)
      # Fit the glm model
      model_xra <- glm(disease_status ~ xra_100 + age, family = binomial, data = data_subset)
      # Construct the file path using accession name and tissue type
      model_path <- file.path(dir_path, paste0(accession_name, "_", .x, "_xra_glm.rds"))
      # Save the model
      saveRDS(model_xra, file = model_path)
      cat("Model saved for", .x, "at", model_path, "\n")
    })
}

# Variables for model saving
main_dir <- "C:/Users/aegea/Documents/TFM/Alzheimer/XRa_models"
accession_name <- "GSE134379"  # Set to the relevant accession number

# Create models for each tissue type and save them
create_glm_models(pheno_xra, main_dir, accession_name)

```
